app/models/product_importer.rb

```diff
@@ -0,0 +1,314 @@
+    else
+      super
+    end
+  end
+
+  private
+
+  def open_spreadsheet
+    if accepted_mimetype
+      @sheet = Roo::Spreadsheet.open(@file, extension: accepted_mimetype)
+    else
+      self.errors.add(:importer, "could not proccess file: invalid filetype")
+      nil
+    end
+  end
+
+  def accepted_mimetype
+    File.extname(@file.path).in?(['.csv', '.xls', '.xlsx', '.ods']) ? @file.path.split('.').last.to_sym : false
+
+    # case @file.content_type
+    #   when "text/csv"
+    #     :csv
+    #   when "application/excel", "application/x-excel", "application/x-msexcel", "application/vnd.ms-excel"
+    #     :xls
+    #   when "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
+    #     :xlsx
+    #   when "application/vnd.oasis.opendocument.spreadsheet"
+    #     :ods
+    #   else
+    #     #Mimetype not compatible
+    #     false
+    # end
+  end
+
+  def headers
+    @sheet.row(1)
+  end
+
+  def rows
+    (2..@sheet.last_row).map do |i|
+      @sheet.row(i)
+    end
+  end
+
+  def entries
+    rows.map do |row|
+      Hash[[headers, row].transpose]
+    end
+  end
+
+  def validate_all
+    suppliers_index = @suppliers_index || get_suppliers_index
+    categories_index = @categories_index || get_categories_index
+
+    entries.each_with_index do |entry, i|
+
+      # Fetch/assign and validate supplier id
+      supplier_name = entry['supplier']
+      if supplier_name.blank?
+        invalidate({i+2 => {entry: entry, errors: ["Supplier name field is empty"]}}) #unless entry['supplier_id']
+      else
+        if suppliers_index[supplier_name]
+          entry['supplier_id'] = suppliers_index[supplier_name]
+        else
+          invalidate({i+2 => {entry: entry, errors: ["Supplier \"#{supplier_name}\" not found in database"]}})
+          #next
+        end
+
+        # Check enterprise permissions
+        unless @editable_enterprises[supplier_name]
+          invalidate({i+2 => {entry: entry, errors: ["You do not have permission to manage products for \"#{supplier_name}\""]}})
+          #next
+        end
+      end
+
+      # Fetch/assign and validate category id
+      category_name = entry['category']
+      if category_name.blank?
+        invalidate({i+2 => {entry: entry, errors: ["Category field is empty"]}}) #unless entry['primary_taxon_id']
+      else
+        if categories_index[category_name]
+          entry['primary_taxon_id'] = categories_index[category_name]
+        else
+          invalidate({i+2 => {entry: entry, errors: ["Category \"#{category_name}\" not found in database"]}})
+          #next
+        end
+      end
+
+      # Ensure on_hand isn't nil (because Spree::Product and Spree::Variant each validate this differently)
+      entry['on_hand'] = 0 if entry['on_hand'].nil?
+
+      # Check if entry can be updated/saved; assign updatable status
+      set_update_status(entry, i)
+
+      # Add valid entry
+      @valid_entries[i+2] = {entry: entry} unless @invalid_entries[i+2]
```

Comment:
Yes, much more readable. Another option to make it even more readable IMHO would be using "early returns". For instance:
```ruby
def mark_as_valid(row_number, entry)
  return if entry_invalid?(row_number)

  add_valid_entry(row_number, entry)
end
```